From 05e69c7042de82d8c37c5e9539fe8669c3bff30c Mon Sep 17 00:00:00 2001
From: Pontus Fuchs <pontus.fuchs@gmail.com>
Date: Fri, 7 Sep 2012 21:30:22 +0200
Subject: [PATCH 21/82] Set wlan mode correctly

FW wants to know if mode is g or b. Find out and tell FW.
---
 ar5523.c |   44 +++++++++++++++++++++++++++++++++++++++-----
 ar5523.h |    2 +-
 2 files changed, 40 insertions(+), 6 deletions(-)

--- a/drivers/net/wireless/ath/ar5523/ar5523.c
+++ b/drivers/net/wireless/ath/ar5523/ar5523.c
@@ -1167,6 +1167,42 @@ static int ar5523_hwconfig(struct ieee80
 	return 0;
 }
 
+static int ar5523_get_wlan_mode(struct ar5523 *ar,
+				 struct ieee80211_bss_conf *bss_conf)
+{
+	struct ieee80211_supported_band *band;
+	int bit;
+	struct ieee80211_sta *sta;
+	u32 sta_rate_set;
+
+	band = ar->hw->wiphy->bands[ar->hw->conf.channel->band];
+	sta = ieee80211_find_sta(ar->vif, bss_conf->bssid);
+	if (!sta) {
+		ar5523_info(ar, "STA not found!\n");
+		return WLAN_MODE_11b;
+	}
+	sta_rate_set = sta->supp_rates[ar->hw->conf.channel->band];
+
+	for (bit = 0; bit < band->n_bitrates; bit++) {
+		if (sta_rate_set & 1) {
+			int rate = band->bitrates[bit].bitrate;
+			switch (rate) {
+			case 60:
+			case 90:
+			case 120:
+			case 180:
+			case 240:
+			case 360:
+			case 480:
+			case 540:
+				return WLAN_MODE_11g;
+			}
+		}
+		sta_rate_set >>= 1;
+	}
+	return WLAN_MODE_11b;
+}
+
 static void ar5523_create_rateset(struct ar5523 *ar,
 				  struct ieee80211_vif *vif,
 				  struct ieee80211_bss_conf *bss_conf,
@@ -1178,7 +1214,6 @@ static void ar5523_create_rateset(struct
 	int bit, i = 0;
 	u32 sta_rate_set;
 
-
 	if (basic)
 		sta_rate_set = bss_conf->basic_rates;
 	else {
@@ -1225,6 +1260,7 @@ static int ar5523_create_connection(stru
 				     struct ieee80211_bss_conf *bss)
 {
 	struct ar5523_cmd_create_connection create;
+	int wlan_mode;
 
 	memset(&create, 0, sizeof(create));
 	create.connid = cpu_to_be32(2);
@@ -1234,10 +1270,8 @@ static int ar5523_create_connection(stru
 
 	ar5523_create_rateset(ar, vif, bss, &create.connattr.rateset, false);
 
-	if (1) //TODO
-		create.connattr.wlanmode = cpu_to_be32(WLAN_MODE_11g);
-	else
-		create.connattr.wlanmode = cpu_to_be32(WLAN_MODE_11b);
+	wlan_mode = ar5523_get_wlan_mode(ar, bss);
+	create.connattr.wlanmode = cpu_to_be32(wlan_mode);
 
 	return ar5523_cmd_write(ar, WDCMSG_CREATE_CONNECTION, &create,
 				sizeof(create), 0);
--- a/drivers/net/wireless/ath/ar5523/ar5523.h
+++ b/drivers/net/wireless/ath/ar5523/ar5523.h
@@ -120,7 +120,7 @@ struct ar5523_cmd_hdr {
 #define	WDCMSG_SET_DEFAULT_KEY		0x43
 
 
-/* OLD commands: REMOVE! */
+/* TODO: OLD commands: REMOVE! */
 #define AR5523_CMD_SET_QUEUE	0x3a
 #define AR5523_CMD_RESET_QUEUE	0x3b
 
